File: slack/t221/verify_family.c
/* verify_family.c — INDEPENDENT enumerator for the "orbits of a base group + prescribed half-turn pairs" families,
 * written from scratch to validate the solver's sym=2+PAIRS (even n, two-loop) and sym=9 (odd n) modes.
 * Configuration = union of whole base-group orbits together with the given half-turn pairs; 2n points;
 * two per row and per column; no three collinear (checked incrementally by forbidden-cell masks).
 * usage: verify_family n base pairspec      base: 4 = quarter turn (C4), 2 = half turn
 *        pairspec like "1,1;4,31"  (each entry c is one cell; its half-turn image is added automatically) */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
typedef unsigned long long u64;
static int n,m,np_pairs;
static u64 forb[40][64];
static int px[128],py[128],np;
static int rowc[64],colc[64];
static long long sols,nodes;
static int orb[4096][4], norb;
static int gcd_(int a,int b){while(b){int t=a%b;a=b;b=t;}return a<0?-a:a;}
static void mark_line(int L,int x1,int y1,int x2,int y2){
    int dx=x2-x1,dy=y2-y1,g=gcd_(dx,dy); if(!g)return; dx/=g; dy/=g;
    int x=x1,y=y1; while(x>=0&&x<n&&y>=0&&y<n){forb[L][x]|=1ULL<<y;x-=dx;y-=dy;}
    x=x1+dx;y=y1+dy; while(x>=0&&x<n&&y>=0&&y<n){forb[L][x]|=1ULL<<y;x+=dx;y+=dy;}
}
static int place(int L,int x,int y){
    if(forb[L][x]>>y&1ULL) return 0;
    for(int i=0;i<np;i++) mark_line(L,px[i],py[i],x,y);
    px[np]=x;py[np]=y;np++; return 1;
}
static int need_orbits;
static char usedorb[4096];
static int touches[64][4096], ntouch[64];
static void rec(int r,int start,int lvl){
    nodes++;
    if(r==n){ sols++; return; }
    if(rowc[r]==2){ rec(r+1,0,lvl); return; }
    for(int i=start;i<ntouch[r];i++){
        int o=touches[r][i];
        if(usedorb[o]) continue;
        int sr[4],sc[4],ok=1;
        for(int t=0;t<4;t++){int x=orb[o][t]/n,y=orb[o][t]%n; sr[t]=x; sc[t]=y; if(++rowc[x]>2||++colc[y]>2) ok=0;}
        if(ok){
            int savednp=np; memcpy(forb[lvl+1],forb[lvl],sizeof(u64)*n);
            int good=1;
            for(int t=0;t<4&&good;t++) good=place(lvl+1,orb[o][t]/n,orb[o][t]%n);
            if(good){ usedorb[o]=1; rec(r,i+1,lvl+1); usedorb[o]=0; }
            np=savednp;
        }
        for(int t=0;t<4;t++){rowc[sr[t]]--; colc[sc[t]]--;}
    }
}
int main(int argc,char**argv){
    if(argc<4){fprintf(stderr,"usage: verify_family n base pairspec\n");return 64;}
    n=atoi(argv[1]); int base=atoi(argv[2]); m=n-1;
    int pc[16],npc=0; { char buf[512]; strncpy(buf,argv[3],511); buf[511]=0;
      for(char*tok=strtok(buf,";");tok;tok=strtok(0,";")){int x,y; if(sscanf(tok,"%d,%d",&x,&y)==2){pc[npc++]=x*n+y; pc[npc++]=(m-x)*n+(m-y);} } }
    char used[4096]; memset(used,0,sizeof(used));
    for(int i=0;i<npc;i++) used[pc[i]]=2;
    /* build orbits of the base group */
    for(int c=0;c<n*n;c++){
        if(used[c]) continue;
        int o[8],k=0,z=c; 
        for(int t=0;t<4;t++){ o[k++]=z; int x=z/n,y=z%n; z = (base==4)? (y*n+(m-x)) : ((m-x)*n+(m-y)); if(z==c) break; }
        int bad=0; for(int t=0;t<k;t++) if(used[o[t]]) bad=1;
        if(bad||k!=4){ for(int t=0;t<k;t++) used[o[t]]=1; continue; }
        for(int t=0;t<k;t++) used[o[t]]=1;
        for(int t=0;t<4;t++) orb[norb][t]=o[t];
        norb++;
    }
    memset(forb,0,sizeof(forb)); memset(rowc,0,sizeof(rowc)); memset(colc,0,sizeof(colc)); np=0;
    for(int i=0;i<npc;i++){ int x=pc[i]/n,y=pc[i]%n; rowc[x]++; colc[y]++; if(!place(0,x,y)){printf("n=%d pairs invalid\n",n); return 0;} }
    need_orbits=(2*n-npc)/4;
    if((2*n-npc)%4){ printf("n=%d base=%d: %d points cannot be filled by size-4 orbits\n",n,base,2*n-npc); return 0; }
    for(int o=0;o<norb;o++) for(int t=0;t<4;t++){int r=orb[o][t]/n; int dup=0; for(int j=0;j<ntouch[r];j++) if(touches[r][j]==o) dup=1; if(!dup) touches[r][ntouch[r]++]=o;}
    rec(0,0,0);
    printf("n=%d base=%d pairs=%s: orbits=%d need=%d INDEPENDENT solutions=%lld nodes=%lld\n",n,base,argv[3],norb,need_orbits,sols,nodes);
    return 0;
}

File: slack/t221/verify_family.py
"""verify_family.py — INDEPENDENT enumeration of one 3-cycle-family subclass, to validate the solver's sym=9/10 modes.
Family (odd n): S = (union of base-group orbits) + the given half-turn pairs; |S| = 2n; two points per row and per column;
no three collinear.  base = C4 (sym 9) or the 'both swaps' group V2/dia2 (sym 10).
usage: python3 verify_family.py n base PAIRS   e.g.  python3 verify_family.py 13 9 "0,1;11,2;10,0" """
import sys
from itertools import combinations

def run(n, base, pairspec):
    m=n-1
    def rot90(c): x,y=divmod(c,n); return y*n+(m-x)
    def half(c):  x,y=divmod(c,n); return (m-x)*n+(m-y)
    def tr(c):    x,y=divmod(c,n); return y*n+x
    def atr(c):   x,y=divmod(c,n); return (m-y)*n+(m-x)
    if base==9:   gen=[rot90]
    elif base==10: gen=[tr,atr]
    else: raise SystemExit("base must be 9 or 10")
    def orbit(c):
        o={c}
        while True:
            new={g(z) for z in o for g in gen}|o
            if new==o: return frozenset(o)
            o=new
    pairs=[]
    for tok in pairspec.split(';'):
        x,y=map(int,tok.split(',')); c=x*n+y; pairs.append((c,half(c)))
    pcells={c for pr in pairs for c in pr}
    if len(pcells)!=2*len(pairs): return None,"degenerate pairs"
    orbs=set()
    for c in range(n*n):
        if c in pcells: continue
        o=orbit(c)
        if o & pcells: continue
        orbs.add(o)
    orbs=[sorted(o) for o in orbs]
    need=2*n-len(pcells)
    orbs=[o for o in orbs if len(o)<=need]
    rowc=[0]*n; colc=[0]*n
    for c in pcells: rowc[c//n]+=1; colc[c%n]+=1
    if max(rowc)>2 or max(colc)>2: return [],"pairs already violate row/col"
    sols=[]; base_pts=sorted(pcells)
    def collinear_ok(P):
        L=len(P)
        for i in range(L):
            x1,y1=divmod(P[i],n)
            for j in range(i+1,L):
                x2,y2=divmod(P[j],n); dx,dy=x2-x1,y2-y1
                for k in range(j+1,L):
                    x3,y3=divmod(P[k],n)
                    if dx*(y3-y1)==dy*(x3-x1): return False
        return True
    def rec(i, chosen, cnt, rc, cc):
        if cnt==need:
            P=sorted(base_pts+[c for o in chosen for c in o])
            if all(v==2 for v in rc) and all(v==2 for v in cc) and collinear_ok(P): sols.append(tuple(P))
            return
        if i>=len(orbs) or cnt>need: return
        # bound: remaining orbits cannot fill
        o=orbs[i]
        if cnt+len(o)<=need:
            ok=True; rc2=rc[:]; cc2=cc[:]
            for c in o:
                rc2[c//n]+=1; cc2[c%n]+=1
                if rc2[c//n]>2 or cc2[c%n]>2: ok=False; break
            if ok: rec(i+1, chosen+[o], cnt+len(o), rc2, cc2)
        rec(i+1, chosen, cnt, rc, cc)
    rec(0, [], 0, rowc, colc)
    return sols, f"orbits available={len(orbs)} need={need} points"

if __name__=='__main__':
    n=int(sys.argv[1]); base=int(sys.argv[2]); spec=sys.argv[3]
    s,info=run(n,base,spec)
    print(f"n={n} base={base} PAIRS={spec}: {info}; INDEPENDENT solutions = {len(s) if s is not None else 'n/a'}")
    for t in (s or [])[:3]: print("   sol", " ".join(map(str,t)))
